File: 0056-merge-intervals/0056-merge-intervals.java
class Solution {
    public int[][] merge(int[][] intervals) {
        List<int[]> list = new ArrayList<>();
        Arrays.sort(intervals, (a, b) -> a[0]!=b[0] ? Integer.compare(a[0], b[0]) : Integer.compare(a[1], b[1]));
        list.add(intervals[0]);
        for(int i=1; i<intervals.length; i++){
            int[] last = list.get(list.size()-1);
            int[] curr = intervals[i];
            //completely disjoint case
            if(last[1] < curr[0]){
                list.add(curr);
            }
            //partially or complete joint case
            else{
                list.set(list.size()-1, new int[]{Math.min(last[0], curr[0]), Math.max(last[1], curr[1])});
            }
        }
        int[][] result = new int[list.size()][2];
        int count = 0;
        for(int[] item: list){
            result[count++] = item;
        }
        return result;
    }
}

File: 0056-merge-intervals/0056-merge-intervals.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals.sort(key = lambda i : i[0])
        res = [[intervals[0][0], intervals[0][1]]]
        for i in range(1, len(intervals)):
            prevEnd = res[-1][1]
            curStart = intervals[i][0]
            if prevEnd >= curStart:
                res.pop()
                res.append([intervals[i - 1][0], max(intervals[i - 1][1], intervals[i][1])])
            else:
                res.append([intervals[i][0], intervals[i][1]])
        return res